File: chalenge/trip/trip.erl
-module (trip).
-export ([minimum_transfer/1]).
-include_lib("eunit/include/eunit.hrl").

init(X) -> lists:sort( fun(A,B) -> A>B end,
  lists:map( fun(N) -> trunc(N*100) end, X) ).
  
to_pay(X) -> 
  Avr = lists:sum(X) div length(X),
  Rem = lists:sum(X) rem length(X),
  lists:duplicate(Rem, Avr + 1) ++ lists:duplicate(length(X) - Rem, Avr).
  
subtract(A, B) -> subtract(A, B, []).
subtract([A|H], [B|T], L) -> subtract(H, T, L ++ [abs(A-B)]);
subtract([], [], L) -> L.

minimum_transfer(X) -> Input = init(X),
  lists:sum(subtract(Input, to_pay(Input))) / 200.0.
  

init_test_() -> [
  ?_assert([3000, 2000, 1000] == init([10.00, 20.00, 30.00]))
].

to_pay_test_() -> [
  ?_assert([2000, 2000, 2000] == to_pay([1000, 2000, 3000])),
  ?_assert([901, 901, 900, 900] == to_pay([1501, 1500, 301, 300]))
].

subtract_test_() -> [
  ?_assert([1,2,3] == subtract([5,4,3], [4,6,6]))
].

minimun_transfer_test_() -> [
  ?_assert(10.00 == minimum_transfer([10.00, 20.00, 30.00])),
  ?_assert(11.99 == minimum_transfer([15.00, 15.01, 3.00, 3.01]))  
].

File: chalenge/trip/trip.rb
%w(rubygems facets/enumerable).each{|lib| require lib}

def minimum_transfer(*vals)
  vals = vals.map{|v| (v*100).to_i }.sort{|x, y| y <=> x}  
  len = vals.length
  
  to_pay = [vals.sum / len] * len             # 공평하게 나누고
  (vals.sum % len).times{|i| to_pay[i] += 1 } # 나머지는 이미 많이 지불한 사람들이 부담
  
  vals.map_with_index{|v, i| (v - to_pay[i]).abs}.sum / 200.0
end

describe "The Trip" do
  it "모든 사람이 사용한 금액이 똑같아지기 위해 전달되어야하는 금액의 총합을 구한다" do
    minimum_transfer(10.00, 20.00, 30.00).should      == 10.00
    minimum_transfer(15.00, 15.01, 3.00, 3.01).should == 11.99
  end
end
